File: frontend/app/modules/ble-audio/src/ble-audioModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import type {
  AudioDeviceSummary,
  BleAudioModuleEvents,
  BufferedAudioExport,
  BufferedAudioStatus,
  HearingSupportStatus,
} from './ble-audio.types';

declare class BleAudioModule extends NativeModule<BleAudioModuleEvents> {
  getStatusAsync(): Promise<HearingSupportStatus>;
  getInputDevicesAsync(): Promise<AudioDeviceSummary[]>;
  getBufferedAudioStatusAsync(): Promise<BufferedAudioStatus>;
  exportBufferedAudioAsync(): Promise<BufferedAudioExport>;
  clearBufferedAudioAsync(): Promise<void>;
  startAsync(configJson: string): Promise<HearingSupportStatus>;
  updateProfileAsync(configJson: string): Promise<HearingSupportStatus>;
  stopAsync(): Promise<HearingSupportStatus>;
}

export default requireNativeModule<BleAudioModule>('ClearHearAudio');

File: frontend/app/modules/ble-audio/android/src/main/java/bleaudio/BleAudioModule.kt
package bleaudio

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class BleAudioModule : Module() {
  private var engine: BleAudioEngine? = null

  override fun definition() = ModuleDefinition {
    Name("ClearHearAudio")

    Events("onStateChange")

    AsyncFunction("getStatusAsync") {
      requireEngine().getStatus()
    }

    AsyncFunction("getInputDevicesAsync") {
      requireEngine().getInputDevices()
    }

    AsyncFunction("getBufferedAudioStatusAsync") {
      requireEngine().getBufferedAudioStatus()
    }

    AsyncFunction("exportBufferedAudioAsync") {
      requireEngine().exportBufferedAudio()
    }

    AsyncFunction("clearBufferedAudioAsync") {
      requireEngine().clearBufferedAudio()
    }

    AsyncFunction("startAsync") { configJson: String ->
      requireEngine().start(configJson)
    }

    AsyncFunction("updateProfileAsync") { configJson: String ->
      requireEngine().updateProfile(configJson)
    }

    AsyncFunction("stopAsync") {
      requireEngine().stop()
    }

    OnDestroy {
      engine?.close()
      engine = null
    }
  }

  private fun requireEngine(): BleAudioEngine {
    engine?.let {
      return it
    }

    val context = appContext.reactContext?.applicationContext
      ?: throw IllegalStateException("React context is not available.")

    return BleAudioEngine(context) { status ->
      sendEvent("onStateChange", status)
    }.also {
      engine = it
    }
  }
}
